### PR TITLE
[Http] Update enrich callbacks for http

### PR DESCRIPTION
**Pull request by vishweshbankwar**
Towards #3495 

## Changes

The PR splits the enrich callback in to three different callbacks based on the object that can be used to enrich the activity.

**Current state:**
### HttpClientInstrumentationOptions
```csharp
using System.Net.Http;

var tracerProvider = Sdk.CreateTracerProviderBuilder()
    .AddHttpClientInstrumentation((options) => options.Enrich
    = (activity, eventName, rawObject) =>
    {
        if (eventName.Equals("OnStartActivity"))
        {
            if (rawObject is HttpRequestMessage request)
            {
                activity.SetTag("requestVersion", request.Version);
            }
        }
        else if (eventName.Equals("OnStopActivity"))
        {
            if (rawObject is HttpResponseMessage response)
            {
                activity.SetTag("responseVersion", response.Version);
            }
        }
        else if (eventName.Equals("OnException"))
        {
            if (rawObject is Exception exception)
            {
                activity.SetTag("stackTrace", exception.StackTrace);
            }
        }
    }).Build();
```

### HttpWebRequestInstrumentationOptions
```csharp
using System.Net;

var tracerProvider = Sdk.CreateTracerProviderBuilder()
    .AddHttpClientInstrumentation((options) => options.Enrich
    = (activity, eventName, rawObject) =>
    {
        if (eventName.Equals("OnStartActivity"))
        {
            if (rawObject is HttpWebRequest request)
            {
                activity.SetTag("requestVersion", request.ProtocolVersion);
            }
        }
        else if (eventName.Equals("OnStopActivity"))
        {
            if (rawObject is HttpWebResponse response)
            {
                activity.SetTag("responseVersion", response.ProtocolVersion);
            }
        }
        else if (eventName.Equals("OnException"))
        {
            if (rawObject is Exception exception)
            {
                activity.SetTag("stackTrace", exception.StackTrace);
            }
        }
    }).Build();
```
**New state:**
#### HttpClientInstrumentationOptions
```csharp
using System.Net.Http;

var tracerProvider = Sdk.CreateTracerProviderBuilder()
    .AddHttpClientInstrumentation((options) =>
    {
        options.EnrichWithHttpRequestMessage = (activity, httpRequestMessage) =>
        {
            activity.SetTag("requestVersion", httpRequestMessage.Version);
        };
        options.EnrichWithHttpResponseMessage = (activity, httpResponseMessage) =>
        {
            activity.SetTag("responseVersion", httpResponseMessage.Version);
        };
        options.EnrichWithException = (activity, exception) =>
        {
            activity.SetTag("stackTrace", exception.StackTrace);
        };
    })
    .Build();
```

#### HttpWebRequestInstrumentationOptions
```csharp
using System.Net;

var tracerProvider = Sdk.CreateTracerProviderBuilder()
    .AddHttpClientInstrumentation((options) =>
    {
        options.EnrichWithHttpWebRequest = (activity, httpWebRequest) =>
        {
            activity.SetTag("requestVersion", httpWebRequest.Version);
        };
        options.EnrichWithHttpWebResponse = (activity, httpWebResponse) =>
        {
            activity.SetTag("responseVersion", httpWebResponse.Version);
        };
        options.EnrichWithException = (activity, exception) =>
        {
            activity.SetTag("stackTrace", exception.StackTrace);
        };
    })
    .Build();
```

For significant contributions please make sure you have completed the following items:

* [x] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

### CODE REVIEW

**Comment by JWilh** [comment]
I personally like it, but maybe it's better to not do a breaking change? You could keep the old Enrich property (marking it as Obsolete only) and simply call both for now and remove it in a "v2" or is this lib still in "prerelease" mode?

**Comment by Kielek** [comment]
> I personally like it, but maybe it's better to not do a breaking change? You could keep the old Enrich property (marking it as Obsolete only) and simply call both for now and remove it in a "v2" or is this lib still in "prerelease" mode?

Prerelease channel, there is not stable releases on the nuget.org: https://www.nuget.org/packages/OpenTelemetry.Instrumentation.Http/#versions-body-tab

**Review by CodeBlanch — APPROVED** [review]
LGTM

File: src/OpenTelemetry.Instrumentation.Http/HttpClientInstrumentationOptions.cs [hunk 1]
```diff
@@ -41,10 +41,32 @@ public class HttpClientInstrumentationOptions
...
         /// </remarks>
-        public Action<Activity, string, object> Enrich { get; set; }
+        public Action<Activity, HttpRequestMessage> EnrichWithHttpRequestMessage { get; set; }
+
+        /// <summary>
+        /// Gets or sets an action to enrich an Activity.
+        /// </summary>
+        /// <remarks>
+        /// <para><see cref="Activity"/>: the activity being enriched.</para>
+        /// <para>string: the name of the event.</para>
```
Comment (utpilla) [inline src/OpenTelemetry.Instrumentation.Http/HttpClientInstrumentationOptions.cs:54]: We can remove this.

File: src/OpenTelemetry.Instrumentation.Http/HttpWebRequestInstrumentationOptions.netfx.cs [hunk 2]
```diff
@@ -41,10 +41,32 @@ public class HttpWebRequestInstrumentationOptions
...
         /// </remarks>
-        public Action<Activity, string, object> Enrich { get; set; }
+        public Action<Activity, HttpWebRequest> EnrichWithHttpWebRequest { get; set; }
+
+        /// <summary>
+        /// Gets or sets an action to enrich an Activity.
+        /// </summary>
+        /// <remarks>
+        /// <para><see cref="Activity"/>: the activity being enriched.</para>
+        /// <para>string: the name of the event.</para>
```
Comment (utpilla) [inline src/OpenTelemetry.Instrumentation.Http/HttpWebRequestInstrumentationOptions.netfx.cs:54]: Same with HttpWebRequest options.

File: src/OpenTelemetry.Instrumentation.Http/HttpClientInstrumentationOptions.cs [hunk 3]
```diff
@@ -41,10 +41,32 @@ public class HttpClientInstrumentationOptions
...
         /// <para>string: the name of the event.</para>
-        /// <para>object: the raw object from which additional information can be extracted to enrich the activity.
+        /// <para><see cref="HttpRequestMessage"/> object from which additional information can be extracted to enrich the activity.
         /// The type of this object depends on the event, which is given by the above parameter.</para>
         /// </remarks>
-        public Action<Activity, string, object> Enrich { get; set; }
+        public Action<Activity, HttpRequestMessage> EnrichWithHttpRequestMessage { get; set; }
+
+        /// <summary>
+        /// Gets or sets an action to enrich an Activity.
```
Comment (utpilla) [inline src/OpenTelemetry.Instrumentation.Http/HttpClientInstrumentationOptions.cs:50]: We should probably mention the raw object type name in the summary description as well.

**Pull request merged**